tools/firmware-utils/src/tplink-safeloader.c

```diff
@@ -683,6 +697,19 @@ static struct image_partition_entry make_soft_version(uint32_t rev) {
 	return entry;
 }
 
+static struct image_partition_entry make_soft_version_from_string(const char *soft_ver) {
+	uint32_t ver_len = strlen(soft_ver);
+	struct image_partition_entry entry = alloc_image_partition("soft-version", ver_len + 2*sizeof(uint32_t));
+
+	uint32_t *len = (uint32_t *)entry.data;
+	*len = htonl(ver_len);
+	*(len+1) = 0;
+	memcpy(len+2, soft_ver, ver_len);
```

Comment:
I think it would be preferable to make the format of the soft-version partition the same as used by the stock firmware, which terminates the string a bit differently: the partition ends with two zero bytes after the newline, the first of which is included in the string length in the partition head; the second is not included in the string length, but in the partition size.